Java and Scala compilers treat records differently, Scala compiler not exposing accessor methods

## Compiler version
3.3.1

## Minimized code
MyRecord.java: 
```
public record MyJavaRecord(String name, Integer age) {}
```
MyTest.scala
```
object MyScalaObject {
  @main def test(): Unit = {
    val javaRecord = new MyJavaRecord("Name", 50)
    println(javaRecord.name())
    println(javaRecord.age)
  }
}
```

If MyJavaRecord is compiled with java compiler, and MyScalaObject is compiled with scala compiler using class file generated by Java compiler for the record, you'd get no syntax errors from scala compiler.

If MyJavaRecord & MyScalaObject are both compiled with scala compiler, first println would give you syntax error.


## Expectation
When record is compiled via Scala compiler, it exposes accessor methods to the fields just like Java compiler.  We should get no errors compiling both with scala compiler.  Scala compiler should not enforce a syntax for java data type that is illegal by java compiler.

Same in Scala 2. Except more carets in Scala 3.
```
➜  i19386 scalac -d /tmp/sandbox *a
MyScalaObject.scala:5: error: not enough arguments for method apply: (i: Int): Char in class StringOps.
Unspecified value parameter i.
    println(javaRecord.name())
                           ^
1 error
➜  i19386 ~/projects/dotty/bin/scalac -d /tmp/sandbox *a
-- [E171] Type Error: MyScalaObject.scala:5:27 -------------------------------------------------------------------------
5 |    println(javaRecord.name())
  |            ^^^^^^^^^^^^^^^^^
  |            missing argument for parameter i of method apply in class StringOps: (i: Int): Char
1 error found
```

--- a/compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala
@@ -860,7 +860,7 @@ object JavaParsers {
 
       // generate the canonical constructor
       val canonicalConstructor =
-        DefDef(nme.CONSTRUCTOR, joinParams(tparams, List(header)), TypeTree(), EmptyTree)
+        DefDef(nme.CONSTRUCTOR, joinParams(Nil, List(header)), TypeTree(), EmptyTree)
           .withMods(Modifiers(Flags.JavaDefined | Flags.Synthetic, mods.privateWithin))
 
       // return the trees
@@ -872,7 +872,7 @@ object JavaParsers {
             tparams = tparams,
             needsDummyConstr = true
           )
-        ).withMods(mods)
+        ).withMods(mods.withFlags(Flags.JavaDefined | Flags.Final))
       }
       addCompanionObject(statics, recordTypeDef)
     end recordDecl

--- a/compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/JavaParsers.scala
@@ -854,7 +854,7 @@ object JavaParsers {
 
       val accessors =
         (for (name, (tpt, annots)) <- fieldsByName yield
-          DefDef(name, Nil, tpt, unimplementedExpr)
+          DefDef(name, List(Nil), tpt, unimplementedExpr)
             .withMods(Modifiers(Flags.JavaDefined | Flags.Method | Flags.Synthetic))
         ).toList
 

--- a/compiler/src/dotty/tools/dotc/typer/Namer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Namer.scala
@@ -906,9 +906,9 @@ class Namer { typer: Typer =>
             && (definesMember || inheritsConcreteMember)
           )
           ||
-          // remove synthetic constructor of a java Record if it clashes with a non-synthetic constructor
-          (denot.isConstructor
-            && isJavaRecord(denot.owner)
+          // remove synthetic constructor or method of a java Record if it clashes with a non-synthetic constructor
+          (isJavaRecord(denot.owner)
+            && (denot.isConstructor || definesMember)
             && denot.owner.unforcedDecls.lookupAll(denot.name).exists(c => c != denot.symbol && c.info.matches(denot.info))
           )
         )

--- a/compiler/src/dotty/tools/dotc/typer/Namer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Namer.scala
@@ -908,7 +908,7 @@ class Namer { typer: Typer =>
           ||
           // remove synthetic constructor or method of a java Record if it clashes with a non-synthetic constructor
           (isJavaRecord(denot.owner)
-            && (denot.isConstructor || definesMember)
+            && denot.is(Method)
             && denot.owner.unforcedDecls.lookupAll(denot.name).exists(c => c != denot.symbol && c.info.matches(denot.info))
           )
         )

--- a/compiler/src/dotty/tools/dotc/typer/Namer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Namer.scala
@@ -879,6 +879,10 @@ class Namer { typer: Typer =>
      *  with a user-defined method in the same scope with a matching type.
      */
     private def invalidateIfClashingSynthetic(denot: SymDenotation): Unit =
+
+      def isJavaRecord(owner: Symbol) =
+        owner.is(JavaDefined) && owner.derivesFrom(defn.JavaRecordClass)
+
       def isCaseClassOrCompanion(owner: Symbol) =
         owner.isClass && {
           if (owner.is(Module)) owner.linkedClass.is(CaseClass)
@@ -904,7 +908,7 @@ class Namer { typer: Typer =>
           ||
           // remove synthetic constructor of a java Record if it clashes with a non-synthetic constructor
           (denot.isConstructor
-            && denot.owner.is(JavaDefined) && denot.owner.derivesFrom(defn.JavaRecordClass)
+            && isJavaRecord(denot.owner)
             && denot.owner.unforcedDecls.lookupAll(denot.name).exists(c => c != denot.symbol && c.info.matches(denot.info))
           )
         )

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -2548,7 +2548,7 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
     def canBeInvalidated(sym: Symbol): Boolean =
       sym.is(Synthetic)
       && (desugar.isRetractableCaseClassMethodName(sym.name) ||
-         (sym.owner.is(JavaDefined) && sym.owner.derivesFrom(defn.JavaRecordClass)))
+         (sym.owner.is(JavaDefined) && sym.owner.derivesFrom(defn.JavaRecordClass) && sym.is(Method)))
 
     if !sym.info.exists then
       // it's a discarded method (synthetic case class method or synthetic java record constructor or overriden member), drop it

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -2548,10 +2548,10 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
     def canBeInvalidated(sym: Symbol): Boolean =
       sym.is(Synthetic)
       && (desugar.isRetractableCaseClassMethodName(sym.name) ||
-        (sym.isConstructor && sym.owner.derivesFrom(defn.JavaRecordClass)))
+        sym.owner.derivesFrom(defn.JavaRecordClass))
 
     if !sym.info.exists then
-      // it's a discarded method (synthetic case class method or synthetic java record constructor), drop it
+      // it's a discarded method (synthetic case class method or synthetic java record constructor or overriden member), drop it
       assert(canBeInvalidated(sym))
       sym.owner.info.decls.openForMutations.unlink(sym)
       return EmptyTree

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -2548,7 +2548,7 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
     def canBeInvalidated(sym: Symbol): Boolean =
       sym.is(Synthetic)
       && (desugar.isRetractableCaseClassMethodName(sym.name) ||
-        sym.owner.derivesFrom(defn.JavaRecordClass))
+         (sym.owner.is(JavaDefined) && sym.owner.derivesFrom(defn.JavaRecordClass)))
 
     if !sym.info.exists then
       // it's a discarded method (synthetic case class method or synthetic java record constructor or overriden member), drop it

--- a/tests/pos-java16+/i19386/FromScala.scala
+++ b/tests/pos-java16+/i19386/FromScala.scala
@@ -1,0 +1,2 @@
+def test(r: R1): Unit =
+  val i: Int = r.i()

--- a/tests/pos-java16+/i19386/R1.java
+++ b/tests/pos-java16+/i19386/R1.java
@@ -1,0 +1,1 @@
+public record R1(int i) {}

--- a/tests/pos-java16+/java-records/FromScala.scala
+++ b/tests/pos-java16+/java-records/FromScala.scala
@@ -1,4 +1,8 @@
 object C:
+
+  def useR0: Unit =
+    val r = R0()
+
   def useR1: Unit =
     // constructor signature
     val r = R1(123, "hello")

--- a/tests/pos-java16+/java-records/FromScala.scala
+++ b/tests/pos-java16+/java-records/FromScala.scala
@@ -41,3 +41,7 @@ object C:
     val l2: Long = r3.l(43L, 44L)
     // supertype
     val isRecord: java.lang.Record = r3
+
+  def useR4: Unit =
+    val r4 = R4(1)
+    val i: Int = r4.t

--- a/tests/pos-java16+/java-records/R0.java
+++ b/tests/pos-java16+/java-records/R0.java
@@ -1,0 +1,1 @@
+public record R0() {}

--- a/tests/pos-java16+/java-records/R4.java
+++ b/tests/pos-java16+/java-records/R4.java
@@ -1,0 +1,1 @@
+public record R4<T>(T t) {}

